lib/spack/llnl/util/tty/__init__.py

```diff
@@ -156,6 +164,11 @@ def msg(message, *args, **kwargs):
 
 
 def info(message, *args, **kwargs):
+    if _osascript:
+        subprocess.call(['osascript', '-e', 'if app "Terminal" is frontmost '
+                         'then tell app "System Events" to keystroke "u" '
+                         'using command down'])
```

Comment:
Three things:

1. I am not sure that `if app "Terminal" is frontmost` is the best condition here.  What if terminal is in the background while the output is spewing, but you want to `ctrl-up` after the fact?  What if you're in xTerm or iTerm, but Terminal is in front (for whatever reason)?

    What you really want to know is whether this `spack` instance was launched from the Terminal, and you can get that by checking the `TERM_PROGRAM` env var.   In Apple Terminal it is set like this:

    ```
    (spackbook):~$ env | grep TERM_PROGRAM
    TERM_PROGRAM=Apple_Terminal
    ```

    Does iTerm support this too?  In iTerm it is set like this:

    ```
    (spackbook):~$ env | grep TERM_PROGRAM
    TERM_PROGRAM=iTerm.app
    ```

    I think you can just check for those, then have the apple script unconditionally talk to "System Events".

2. The other thing that you likely need to factor in here is whether the output is going to `stdout`.  Have you tried this with `spack install` without `-v`?  I think you should maybe think about that, though maybe just doing it in `msg` and `info` is enough.

3. Are there markers that need to be suppressed with the terminal is not interactive?  What if the user runs this and sends output to a file `spack install -v hdf5 > log` (people do this).  You should probably only print out these markers `sys.stdout.isatty()` returns `True`.

Reply:
The reason for `if app "Terminal" is frontmost` might be that it causes the Terminal to steal focus. I tried asking what the purpose was and got a response at the bottom of https://scriptingosx.com/2017/03/terminal-the-marks-the-spot/. I'll have to play around with this and confirm. I'll see what happens when the output is not verbose or is piped to a log.

Reply:
Some observations:

1. Even if I remove `if app "Terminal" is frontmost`, it still doesn't add mark characters to the Terminal if I'm not currently focused on the Terminal. Luckily, it doesn't steal focus, but it does sound the terminal bell, which is annoying...
2. This doesn't seem to be an issue. I believe the `osascript` command is still run, and you can see that it adds a mark before the next `==>` is displayed, but it doesn't break anything or add duplicate marks. Probably worth checking if verbose though.
3. This doesn't seem to be an issue. No meta characters are added to the log, although it has the same minor side-effect as 2.

I'm going to see if I can find a way to fix 1. It would be great if we could always add a mark character even if the Terminal is in the background. I also want to disable that bell...

Reply:
Posted on StackOverflow: https://stackoverflow.com/questions/56262815/add-mark-to-terminal-in-background-using-applescript

Reply:
I'm pretty sure that if you remove `if app "Terminal" is frontmost then` (i.e. only have `tell app "System Events" to keystroke "u" using command down`) then it will issue the keystroke to whatever app is currently active, causing who knows what. As a side note, I suspect that if one changes the shortcut then the AppleScript won't work—it might be more robust to script clicking the menu instead.

Reply:
Another important test case will be what happens when multiple Terminal windows are open.